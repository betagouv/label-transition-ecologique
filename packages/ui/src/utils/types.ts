import { LinkProps } from 'next/link';
import {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  HTMLAttributes,
} from 'react';

/** Propriétés HTML d'une div */
export type DivHTMLProps = HTMLAttributes<HTMLDivElement>;
/** Propriétés HTML d'un bouton */
export type ButtonHTMLProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  [key: `data-${string}`]: unknown;
};

/** Propriétés HTML d'un lien */
export type AnchorHTMLProps = AnchorHTMLAttributes<HTMLAnchorElement>;

/** Props du Link de Nextjs reconstituée car non exportées entièrement */
export type LinkFullProps = Omit<
  React.AnchorHTMLAttributes<HTMLAnchorElement>,
  keyof LinkProps
> &
  LinkProps & {
    children?: React.ReactNode;
  } & React.RefAttributes<HTMLAnchorElement>;

/** Détermine si les proriétés appartiennent à un lien */
export function isLink(
  props: ButtonHTMLProps | LinkFullProps | DivHTMLProps
): props is LinkFullProps {
  return (props as LinkFullProps).href !== undefined;
}

/** Type générique pour la gestion d'un état d'ouverture  */
export type OpenState = {
  /** état d'ouverture */
  isOpen: boolean;
  /* accompagne "isOpen" afin de pouvoir fermer la modale */
  setIsOpen: (opened: boolean) => void;
};
