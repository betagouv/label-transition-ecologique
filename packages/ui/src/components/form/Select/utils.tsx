import {Option, OptionSection, SelectOption} from './Options';

/** Option section type guards */
export function isOptionSection(option: SelectOption): option is OptionSection {
  return (option as OptionSection).title !== undefined;
}

/** Option type guards */
export function isOption(option: SelectOption): option is Option {
  type NewType = Option;
  return (option as NewType).value !== undefined;
}

/** Renvoie un tableau d'options, quelles soient dans une section ou non */
export const getOptions = (selectOptions: SelectOption[]): Option[] =>
  selectOptions.reduce(
    (acc: Option[], curr) =>
      acc.concat(isOptionSection(curr) ? curr.options : curr),
    []
  );
