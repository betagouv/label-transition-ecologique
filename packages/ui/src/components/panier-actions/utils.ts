export type Complexite = 'simple' | 'intermédaire' | 'élevée';

export type State = 'success' | 'warning' | 'error';

export const valeurToBadge: Record<number, {nom: Complexite; style: State}> = {
  1: {nom: 'simple', style: 'success'},
  2: {nom: 'intermédaire', style: 'warning'},
  3: {nom: 'élevée', style: 'error'},
};
