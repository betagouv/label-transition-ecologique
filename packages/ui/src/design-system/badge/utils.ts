import {BadgeState} from './index';

/**
 * Thème couleurs du composant Badge
 *
 * Pour un variant donné, permet de récupérer les couleurs de
 * texte, background, border et icône en fonction du statut du bouton
 **/
export const badgeClassnames: Record<
  BadgeState,
  {
    text: string;
    background: string;
    border: string;
    icon: string;
  }
> = {
  default: {
    text: 'text-grey-8',
    background: 'bg-grey-1',
    border: 'border-grey-5',
    icon: 'text-primary-7',
  },
  standard: {
    text: 'text-primary-7',
    background: 'bg-primary-2',
    border: 'border-primary-2',
    icon: 'text-primary-7',
  },
  success: {
    text: 'text-success-1',
    background: 'bg-success-2',
    border: 'border-success-2',
    icon: 'text-success-1',
  },
  warning: {
    text: 'text-warning-1',
    background: 'bg-warning-2',
    border: 'border-warning-2',
    icon: 'text-warning-1',
  },
  new: {
    text: 'text-new-1',
    background: 'bg-new-2',
    border: 'border-new-2',
    icon: 'text-new-1',
  },
  error: {
    text: 'text-error-1',
    background: 'bg-error-2',
    border: 'border-error-2',
    icon: 'text-error-1',
  },
  info: {
    text: 'text-info-1',
    background: 'bg-info-2',
    border: 'border-info-2',
    icon: 'text-info-1',
  },
  grey: {
    text: 'text-grey-6',
    background: 'bg-grey-3',
    border: 'border-grey-3',
    icon: 'text-grey-6',
  },
};
