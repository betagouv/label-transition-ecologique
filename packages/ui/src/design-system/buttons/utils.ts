import {ButtonVariant} from './types';

/**
 * Thème couleurs du composant Button
 *
 * Pour un variant donné, permet de récupérer les couleurs de
 * texte, background, border et icône en fonction du statut du bouton
 *
 * Utilisation des couleurs configurées dans Tailwind
 **/
export const buttonThemeClassnames: Record<
  ButtonVariant,
  {
    text: string;
    background: string;
    border: string;
    icon: string;
  }
> = {
  primary: {
    text: 'text-white hover:!text-white disabled:!text-primary-3',
    background: 'bg-primary-7 hover:!bg-primary-8 disabled:!bg-primary-4',
    border:
      'border-primary-7 hover:!border-primary-8 disabled:!border-primary-4',
    icon: 'fill-white group-hover:fill-white group-disabled:fill-primary-3',
  },
  secondary: {
    text: 'text-white hover:!text-white disabled:!text-grey-1',
    background: 'bg-secondary-1 hover:!bg-secondary-2 disabled:!bg-secondary-2',
    border:
      'border-secondary-1 hover:!border-secondary-2 disabled:!border-secondary-2',
    icon: 'fill-white group-hover:fill-white group-disabled:fill-grey-1',
  },
  outlined: {
    text: 'text-primary-7 hover:!text-primary-8 disabled:!text-primary-3',
    background: 'bg-white hover:!bg-primary-1 disabled:!bg-white',
    border:
      'border-primary-7 hover:!border-primary-8 disabled:!border-primary-4',
    icon: 'fill-primary-7 group-hover:fill-primary-8 group-disabled:fill-primary-3',
  },
  white: {
    text: 'text-primary-7 hover:!text-primary-8 disabled:!text-primary-4',
    background: 'bg-white hover:!bg-primary-1 disabled:!bg-white',
    border: 'border-white hover:!border-primary-1 disabled:!border-white',
    icon: 'fill-primary-7 group-hover:fill-primary-8 group-disabled:fill-primary-4',
  },
  grey: {
    text: 'text-primary-7 hover:!text-primary-8 disabled:!text-primary-4',
    background: 'bg-grey-1 hover:!bg-grey-2 disabled:!bg-white',
    border: 'border-grey-4 hover:!border-grey-4 disabled:!border-grey-2',
    icon: 'fill-primary-7 group-hover:fill-primary-8 group-disabled:fill-primary-4',
  },
};
