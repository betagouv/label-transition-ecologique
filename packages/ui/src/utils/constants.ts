export const BASE_URL = 'https://territoiresentransitions.fr';
export const APP_BASE_URL = 'https://app.territoiresentransitions.fr';
