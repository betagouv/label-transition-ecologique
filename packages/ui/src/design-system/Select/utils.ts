import {naturalSort} from '../../utils/naturalSort';

/**
 * Types partagés entre tous les composants selects
 * (Select, MultiSelect, MultiSelectFilter)
 */

/** Type du champ valeur d'une option */
export type OptionValue = number | string;
/** Type de base d'une option générique */
export type Option = {value: OptionValue; label: string};
/** Type d'une liste d'options dans un sélecteur */
export type OptionSection = {
  title: string;
  options: Option[];
};
/** Type d'une option dans un sélecteur, peut être une simple option ou une liste d'options */
export type SelectOption = Option | OptionSection;

/** Option section type guards */
export function isOptionSection(option: SelectOption): option is OptionSection {
  return (option as OptionSection).title !== undefined;
}

/** Option type guards */
export function isSingleOption(option: SelectOption): option is Option {
  type NewType = Option;
  return (option as NewType).value !== undefined;
}

/** Renvoie un tableau d'options, quelles soient dans une section ou non */
export const getFlatOptions = (selectOptions: SelectOption[]): Option[] =>
  selectOptions.reduce(
    (acc: Option[], curr) =>
      acc.concat(isOptionSection(curr) ? curr.options : curr),
    []
  );

/** Extrait le label d'une option dans une liste d'options */
export const getOptionLabel = (optionValue: OptionValue, options: Option[]) =>
  options.find((v: Option) => v.value?.toString() === optionValue?.toString())
    ?.label;

/** Tri de façon naturelle (1, 2, 11, 40, 'a', 'b') une liste d'options par ordre alpha-numérique */
export const sortOptionByAlphabet = (
  options: SelectOption[]
): SelectOption[] => {
  // crée 2 tableaux pour les options et les section afin de leur appliquer le sort
  const optionArray: Option[] = [];
  const sectionArray: OptionSection[] = [];

  options.forEach(option => {
    if (isSingleOption(option)) {
      optionArray.push(option);
    } else {
      sectionArray.push(option);
    }
  });

  // sort options
  optionArray.sort((a, b) =>
    naturalSort(a.label.toUpperCase(), b.label.toUpperCase())
  );

  // sort sections
  sectionArray.forEach(section => {
    section.options.sort((a, b) =>
      naturalSort(a.label.toUpperCase(), b.label.toUpperCase())
    );
  });

  return [...optionArray, ...sectionArray];
};

/**
 * Filtre les options, quelles soient dans une section ou non.
 * Utilisée dans les sélecteurs avec saisie.
 * Renvoi une liste d'options ou des sections avec les options filtrées
 */
export const filterOptions = (
  options: SelectOption[],
  filterValue: string
): SelectOption[] =>
  options.reduce((acc: SelectOption[], currentOption) => {
    if (isSingleOption(currentOption)) {
      if (
        currentOption.label.toLowerCase().includes(filterValue.toLowerCase())
      ) {
        return [...acc, currentOption];
      }
    }

    if (isOptionSection(currentOption)) {
      const filteredOptions = currentOption.options.filter(option =>
        option.label.toLowerCase().includes(filterValue.toLowerCase())
      );
      if (filteredOptions.length > 0) {
        return [
          ...acc,
          {
            title: currentOption.title,
            options: filteredOptions,
          },
        ];
      } else {
        return acc;
      }
    }

    return acc;
  }, []);
