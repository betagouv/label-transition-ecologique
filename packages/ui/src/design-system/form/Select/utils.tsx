import {naturalSort} from '../../../utils/naturalSort';
import {Option, OptionSection, OptionValue, SelectOption} from './Options';

/** Option section type guards */
export function isOptionSection(option: SelectOption): option is OptionSection {
  return (option as OptionSection).title !== undefined;
}

/** Option type guards */
export function isSingleOption(option: SelectOption): option is Option {
  type NewType = Option;
  return (option as NewType).value !== undefined;
}

/** Renvoie un tableau d'options, quelles soient dans une section ou non */
export const getFlatOptions = (selectOptions: SelectOption[]): Option[] =>
  selectOptions.reduce(
    (acc: Option[], curr) =>
      acc.concat(isOptionSection(curr) ? curr.options : curr),
    []
  );

/** Extrait le label d'une option dans une liste d'options */
export const getOptionLabel = (optionValue: OptionValue, options: Option[]) =>
  options.find((v: Option) => v.value === optionValue).label;

/** Gère la sélection/désélection d'une valeur simple pour le composant Select */
export const onSelectSingle = (
  optionValue: OptionValue,
  value: OptionValue | undefined
) => {
  // si la valeur de l'option sélectionnée est différente de la valeur du sélecteur
  // (sélection)
  if (optionValue !== value) {
    // alors on renvoie la nouvelle valeur
    return optionValue;
    // sinon on retourne undefined
    // (désélection)
  } else {
    return undefined;
  }
};

/** Tri de façon naturelle (1, 2, 11, 40, 'a', 'b') une liste d'options par ordre alpha-numérique */
export const sortOptionByAlphabet = (
  options: SelectOption[]
): SelectOption[] => {
  // crée 2 tableaux pour les options et les section afin de leur appliquer le sort
  const optionArray: Option[] = [];
  const sectionArray: OptionSection[] = [];

  options.forEach(option => {
    if (isSingleOption(option)) {
      optionArray.push(option);
    } else {
      sectionArray.push(option);
    }
  });

  // sort options
  optionArray.sort((a, b) =>
    naturalSort(a.label.toUpperCase(), b.label.toUpperCase())
  );

  // sort sections
  sectionArray.forEach(section => {
    section.options.sort((a, b) =>
      naturalSort(a.label.toUpperCase(), b.label.toUpperCase())
    );
  });

  return [...optionArray, ...sectionArray];
};

/** Gère la sélection/désélection d'un tableau de valeurs pour le composant Select */
export const onSelectMultiple = (
  optionValue: OptionValue,
  values: OptionValue[] | undefined
) => {
  // si au moins une valeur est présente dans les valeurs du sélecteur
  if (values) {
    if (values.includes(optionValue)) {
      // retrait d'une valeur
      return values.length === 1
        ? // renvoie undefined si la seule valeur présente dans les valeurs du sélecteur est la même que la valeur de l'option
          undefined
        : values.filter(v => v !== optionValue);
    } else {
      // ajoût d'une valeur
      return [...values, optionValue];
    }
    // si aucune valeur n'était déjà sélectionnée alors on renvoie directement la veleur de l'option dans un tableau
  } else {
    return [optionValue];
  }
};

/**
 * Filtre les options, quelles soient dans une section ou non.
 * Utilisée dans les sélecteurs avec saisie.
 * Renvoi une liste d'options ou des sections avec les options filtrées
 */
export const filterOptions = (
  options: SelectOption[],
  filterValue: string
): SelectOption[] =>
  options.reduce((acc: SelectOption[], currentOption) => {
    if (isSingleOption(currentOption)) {
      if (
        currentOption.label.toLowerCase().includes(filterValue.toLowerCase())
      ) {
        return [...acc, currentOption];
      }
    }

    if (isOptionSection(currentOption)) {
      const filteredOptions = currentOption.options.filter(option =>
        option.label.toLowerCase().includes(filterValue.toLowerCase())
      );
      if (filteredOptions.length > 0) {
        return [
          ...acc,
          {
            title: currentOption.title,
            options: filteredOptions,
          },
        ];
      } else {
        return acc;
      }
    }

    return acc;
  }, []);
