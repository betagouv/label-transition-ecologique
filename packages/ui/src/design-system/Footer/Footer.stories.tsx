import {Meta, StoryObj} from '@storybook/react';
import {Footer} from './Footer';
import {APP_BASE_URL, BASE_URL} from 'utils/constants';

const meta: Meta<typeof Footer> = {
  component: Footer,
  args: {
    logos: [<i className="ri-leaf-line text-primary-9 !text-8xl" />],
    content:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam eget maximus nisl, eget ullamcorper mauris. Nulla sagittis tempus leo nec elementum.',
    contentLinks: [
      {
        label: 'territoiresentransitions.fr',
        href: BASE_URL,
        external: true,
      },
    ],
    bottomContent: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
    bottomLinks: [
      {
        label: 'Site',
        href: BASE_URL,
        external: true,
      },
      {
        label: 'Application',
        href: APP_BASE_URL,
        external: true,
      },
    ],
  },
};

export default meta;

type Story = StoryObj<typeof Footer>;

export const Default: Story = {};
