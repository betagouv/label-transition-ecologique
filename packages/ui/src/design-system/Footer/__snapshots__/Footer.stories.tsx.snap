// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Footer Default smoke-test 1`] = `
<footer class="w-full border-t-2 border-t-primary-8">
  <div class="w-full mx-auto px-4 lg:px-6 xl:max-w-7xl xl:px-2 py-8 flex max-md:flex-wrap justify-between md:gap-x-16 lg:gap-x-32 gap-y-6 md:flex-row">
    <div class="flex h-28 gap-x-1">
      <div class="h-full">
        <i class="ri-leaf-line text-primary-9 !text-8xl">
        </i>
      </div>
    </div>
    <div>
      <p class="text-sm leading-6 text-grey-9">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nam eget maximus nisl, eget ullamcorper mauris. Nulla sagittis tempus leo nec elementum.
      </p>
      <ul class="list-none flex flex-wrap gap-4 divide-x divide-grey-4 mb-0">
        <li class="pl-4 first-of-type:pl-0 pb-0">
          <a href="https://territoiresentransitions.fr"
             class="bg-none after:hidden w-fit flex items-center border-solid group gap-1 cursor-pointer border-b border-t-0 border-x-0 !p-px font-medium hover:border-b-2 hover:!pb-0 text-primary-8 !bg-transparent border-primary-7 text-sm py-2.5 px-6 flex-row-reverse !text-primary-8 border-b-transparent hover:border-b-primary-8 hover:!border-b"
             target="_blank"
             rel="noreferrer noopener"
          >
            <span class="ri-external-link-line font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] fill-primary-8 mb-1">
            </span>
            territoiresentransitions.fr
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div class="w-full border-t border-t-primary-4">
    <div class="w-full mx-auto px-4 lg:px-6 xl:max-w-7xl xl:px-2 py-6">
      <ul class="list-none flex flex-wrap gap-2 divide-x divide-grey-4">
        <li class="pl-2 first-of-type:pl-0 pb-0">
          <a href="https://territoiresentransitions.fr"
             class="bg-none after:hidden w-fit flex items-center border-solid group gap-1 cursor-pointer border-b border-t-0 border-x-0 !p-px font-medium hover:border-b-2 hover:!pb-0 text-primary-8 !bg-transparent border-primary-7 text-xs py-2 px-5 flex-row-reverse !text-grey-8 border-b-transparent hover:border-b-grey-8 hover:!border-b !font-normal"
             target="_blank"
             rel="noreferrer noopener"
          >
            <span class="ri-external-link-line font-normal text-center h-3.5 w-3.5 text-[0.875rem] leading-[0.875rem] fill-primary-8 mb-0.5">
            </span>
            Site
          </a>
        </li>
        <li class="pl-2 first-of-type:pl-0 pb-0">
          <a href="https://app.territoiresentransitions.fr"
             class="bg-none after:hidden w-fit flex items-center border-solid group gap-1 cursor-pointer border-b border-t-0 border-x-0 !p-px font-medium hover:border-b-2 hover:!pb-0 text-primary-8 !bg-transparent border-primary-7 text-xs py-2 px-5 flex-row-reverse !text-grey-8 border-b-transparent hover:border-b-grey-8 hover:!border-b !font-normal"
             target="_blank"
             rel="noreferrer noopener"
          >
            <span class="ri-external-link-line font-normal text-center h-3.5 w-3.5 text-[0.875rem] leading-[0.875rem] fill-primary-8 mb-0.5">
            </span>
            Application
          </a>
        </li>
      </ul>
      <div class="text-xs leading-5 text-grey-8">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
      </div>
    </div>
  </div>
</footer>
`;
